### ISSUE TITLE
508-defect-2 [SEMANTIC MARKUP, COGNITION]: Groups of form elements MUST have group labels (e.g. <fieldset> and <legend>)

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va-forms-system-core#314]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:** Josh Kim

## User Story or Problem Statement

As a screen reader user, I need to know that a range of form inputs are grouped together, so that I may understand the scope of the use case. As a user with cognitive considerations, I want to know which page of the form I am editing so that I can update the right information.

## Details

Upon clicking "edit" on any of the pages within the applicant information accordion, the heading element that indicates the page disappears leaving only the form inputs behind. This may cause several issues:

1. Screen reader users will struggle to navigate through the forms as heading elements are actively inserted and removed from the page. It is also not announced to the user
1. Sighted users with cognitive considerations may lost track of their place within the multiple forms nested in an accordion, particularly on mobile.
1. Missing heading elements eliminates the opportunity to make use of aria-describedby to provide unique accessible names to the "update page" buttons.

These issues are further compounded by a lack of groupings for the form itself. Groups of form elements MUST have group labels to meet Section 508 requirements.

## Acceptance Criteria

- [ ] Forms belong to logical groupings using fieldsets and legends.
- [ ] Headings do not disappear after clicking "edit."
- [ ] Update page buttons point to headings using aria-describedby.

## Environment

* Operating System: all
* Browser: any
* Screenreading device: any
* Server destination: staging, production

## Steps to Recreate

1. Fill out Form 28-8832 as a spouse/child and reach the Step 2 of 2: Review Application page.
2. Open the "Application Information" accordion.
3. Click "edit" next to Applicant Information.
4. Verify the heading "Application Information" has been removed.
5. Open dev tools and verify no fieldsets and legends exist.
6. Open the "Your sponsoring..." accordion.
7. Verify no heading exists.

## Solution
Add a fieldset around each group along with a legend that contains the formerly deleted heading. This enhancement will provide multiple methods of navigation to screen reader users through headings and form groupings. It will also help sighted users on mobile understand which pages they are editing as they scroll.

It should be noted that this may be a sitewide issue that may need to be addressed in the form/design system itself. Alternative, more optimal patterns should be explored. I strongly recommend considering [gov.uk's check answers pattern](https://design-system.service.gov.uk/patterns/check-answers/) which would alleviate many of the technical issues the existing pattern creates.

❗️ Please reference [the ticket on headings for this page](https://github.com/department-of-veterans-affairs/va.gov-team/issues/16763) before addressing this. This ticket requests that headings be added to the accordions, which consequentially bumps down headings for the rest of the page.

### Before

```html
<!-- Fields -->
<div>...</div>
<button type="submit" class="usa-button-primary null" id="9-continueButton">
  Update page
</button>

<!-- Fields -->
<div>...</div>
<button type="submit" class="usa-button-primary null" id="9-continueButton">
  Update page
</button>

<!-- Fields -->
<div>...</div>
<button type="submit" class="usa-button-primary null" id="9-continueButton">
  Update page
</button>

```

### After

``` diff
+ <fieldset>
+  <legend id="id-1">
+    <h4 class="vads-u-font-size--h5 vads-u-margin--0">
+      Applicant Information 
+    </h4>
+   </legend>
  <!-- Fields -->
  <div>...</div>
+  <!-- Add aria-describedby to this button-->
  <button type="submit" class="usa-button-primary null" id="9-continueButton" aria-describedby="id-1">
    Update page
  </button>
+ </fieldset>

<!-- Repeat changes down the page -->
```

## WCAG or Vendor Guidance (optional)

* [Grouping controls](https://www.w3.org/WAI/tutorials/forms/grouping/)
* [WCAG 2.0 1.3.1 Info and Relationships A](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships)

### Before
Note the missing title for this form group.
<img width="685" alt="Screen Shot 2020-12-01 at 11 07 31 AM" src="https://user-images.githubusercontent.com/14154792/100765357-6b7f2280-33c5-11eb-9329-a75905338f7f.png">

### After
Note the reinstated heading. 
<img width="688" alt="Screen Shot 2020-12-01 at 11 08 01 AM" src="https://user-images.githubusercontent.com/14154792/100765416-7b970200-33c5-11eb-98e6-0bc7cd498aa4.png">

**Comment by taylorkaren** [comment]
Info from @k80bowman: This sounds like a review page issue. I know there are review page re-design things going on, so I'm not sure if we want to tackle this or not. I did not re-test myself, but I'm pretty confident this is still an issue. - Katy

**Comment by caw310** [comment]
@taylorkaren , this was another item we discussed as belonging in the FLT backlog.

**Comment by obliviga** [comment]
The only thing we can feasibly update here IMO is the aria-describedby attribute on the Continue button, which I have created a PR for here: https://github.com/department-of-veterans-affairs/vets-website/pull/21490

The reason we cannot add a legend is because we already have a heading outside of the form that is associated with the form elements:
![image.png](https://images.zenhubusercontent.com/6283eaffdea4c21364d67e5a/ba43b2b7-f92f-4c9a-b2b9-b5ef4ad51f90)

Adding fieldset and legend elements will change how the form looks.

**Comment by taylorkaren** [comment]
@obliviga: Let's get input from @joshkimux on the fieldset and legend inputs. If those still need to be addressed, let's create a new ticket with the specific details around that request.

**Comment by taylorkaren** [comment]
Let's go ahead and put through the PR for the  aria-describedby attribute on the Continue button. We will move Josh's last comment to a separate ticket may address it in Q3 2022 if we work with the Benefits team on piloting the VAFSC.

**Comment by obliviga** [comment]
Merged the PR, now waiting for successful deployment to production.

**Comment by obliviga** [comment]
Verified in production:![image.png](https://images.zenhubusercontent.com/6283eaffdea4c21364d67e5a/b5a9e205-c59c-4d67-93ff-372084187c2a)
